samcli/hook_packages/terraform/hooks/prepare/resource_linking.py

```diff
@@ -99,13 +104,111 @@ def _get_configuration_address(address: str) -> str:
+
+    output = module.outputs.get(output_name)
+
+    if not output:
+        raise InvalidResourceLinkingException(f"Output {output_name} was not found in module {module.full_address}")
+
+    output_value = output.value
+
+    LOG.debug("Resolving output {%s} for module {%s}", output_name, module.full_address)
+
+    if isinstance(output, ConstantValue):
+        LOG.debug(
+            "Resolved constant value {%s} for module {%s} for output {%s}",
+            output.value,
+            module.full_address,
+            output_name,
+        )
+
+        results.append(output)
+    elif isinstance(output, References):
+        LOG.debug("Found references for module {%s} for output {%s}", module.full_address, output_name)
+
+        cleaned_references = _clean_references_list(output_value)
+
+        for reference in cleaned_references:
+            if reference.startswith("var."):
+                LOG.debug(
+                    "Resolving variable reference {%s} for module {%s} for output {%s}",
+                    reference,
+                    module.full_address,
+                    output_name,
+                )
+
+                stripped_reference = _get_configuration_address(reference[reference.find(".") + 1 :])
+                results += _resolve_module_variable(module, stripped_reference)
+            elif reference.startswith("module."):
+                LOG.debug(
+                    "Resolving module reference {%s} for module {%s} for output {%s}",
+                    reference,
+                    module.full_address,
+                    output_name,
+                )
+
+                # module.bbb.ccc => bbb
+                module_name = reference[reference.find(".") + 1 : reference.rfind(".")]
+                # module.bbb.ccc => ccc
+                output_name = reference[reference.rfind(".") + 1 :]
+
+                stripped_reference = _get_configuration_address(module_name)
+
+                if not module.child_modules:
+                    raise InvalidResourceLinkingException(
+                        f"Module {module.full_address} does not have child modules defined, possible misconfiguration"
+                    )
+
+                child_module = module.child_modules.get(stripped_reference)
+
+                if not child_module:
+                    raise InvalidResourceLinkingException(
+                        f"Module {module.full_address} does not have {stripped_reference} as a child module"
+                        ", possible misconfiguration"
+                    )
+
+                results += _resolve_module_output(child_module, output_name)
+            else:
+                LOG.debug(
+                    "Resolved reference {%s} for module {%s} for output {%s}",
+                    reference,
+                    module.full_address,
+                    output_name,
+                )
+
+                results.append(ResolvedReference(reference, module.full_address or ""))
```

Comment:
Haven't reviewed the whole PR but remembered that we discussed this part in the office. Let's make the module address optional instead, and set this as None here (instead of empty string) 🙏